File: scripts/test_solutions/211_add_search_words.js
var WordDictionary = function() { this.root = {}; };
WordDictionary.prototype.addWord = function(word) {
    let node = this.root;
    for (const c of word) { if (!node[c]) node[c] = {}; node = node[c]; }
    node.isEnd = true;
};
WordDictionary.prototype.search = function(word) {
    function dfs(node, i) {
        if (i === word.length) return node.isEnd === true;
        if (word[i] === '.') {
            for (const key in node) { if (key !== 'isEnd' && dfs(node[key], i + 1)) return true; }
            return false;
        }
        return node[word[i]] ? dfs(node[word[i]], i + 1) : false;
    }
    return dfs(this.root, 0);
};

File: scripts/test_solutions/211_add_search_words.cpp
class WordDictionary {
private:
    struct TrieNode {
        unordered_map<char, TrieNode*> children;
        bool isEnd = false;
    };
    TrieNode* root;

    bool dfs(TrieNode* node, const string& word, int i) {
        if (i == word.size()) {
            return node->isEnd;
        }
        if (word[i] == '.') {
            for (auto& p : node->children) {
                if (dfs(p.second, word, i + 1)) {
                    return true;
                }
            }
            return false;
        }
        if (node->children.find(word[i]) == node->children.end()) {
            return false;
        }
        return dfs(node->children[word[i]], word, i + 1);
    }

public:
    WordDictionary() {
        root = new TrieNode();
    }

    void addWord(string word) {
        TrieNode* node = root;
        for (char c : word) {
            if (node->children.find(c) == node->children.end()) {
                node->children[c] = new TrieNode();
            }
            node = node->children[c];
        }
        node->isEnd = true;
    }

    bool search(string word) {
        return dfs(root, word, 0);
    }
};
